File: php/test-mafreebox.php
<?php

error_reporting(E_ALL);

require_once('lib/Mafreebox.php');

function get_arg($idx=1, $default=false){
	if(isset($_SERVER["argv"][$idx]))
		return $_SERVER["argv"][$idx];
	return $default;
		
}

function file_rotate($file, $count, $dir='var/'){

	$infos = pathinfo($file);
	$files = glob(sprintf('%s%s.%s.*',$dir, $infos['filename'],$infos['extension']));
	# sort($files, SORT_NATURAL);
	natcasesort($files);
	$files = array_reverse($files);

	foreach($files as $i => $f){
		if((count($files) -$i) > $count-1){
			printf("## unlinking file $f \n");
			# unlink($f);
		}
		else {
			$base = basename($f);
			$inf = pathinfo($f);
			$id = $inf['extension'];
			$name = sprintf("%s/%s", $inf['dirname'], $inf['filename']);
			$f1 = sprintf("%s.%s", $name, $id);
			$f2 = sprintf("%s.%s", $name, $id+1);
			# printf("## rename %s -> %s\n", $f1, $f2 );
			if(file_exists($f1)) rename($f1, $f2);
		}
	}
	
	$path = sprintf('%s%s',$dir, $file);
	if(file_exists($path)) rename($path, $path.'.1');
}

# mettre les paramêtres de connexion de préférence dans le fichier de configuration 
# valable surtout quand on contribue au source (dépot GIT par exemple).
#
$config_file = sprintf('%s/.config/mafreebox.cfg', getenv('HOME'));
if(file_exists($config_file)){
	require_once($config_file);
} else
	$config = array(
		'url'      => 'http://mafreebox.freebox.fr/',
		'user'     => 'freebox',
		'password' => '123456' # your password
	);

$freebox = new Mafreebox($config['url'], $config['user'], $config['password']);


/*
	period = {hour|day|week}
	type={rate|snr}
	dir={down|up}

*/

function rrd_daily($freebox){
	$f = 'rrd-rate-day-up.png';   file_rotate("$f", 31); file_put_contents("var/$f", $freebox->rrd_graph($type='rate', $period='day', $dir='up'));
	$f = 'rrd-rate-day-down.png'; file_rotate("$f", 31); file_put_contents("var/$f", $freebox->rrd_graph($type='rate', $period='day', $dir='down'));
	$f = 'rrd-snr-day.png';       file_rotate("$f", 31); file_put_contents("var/$f", $freebox->rrd_graph($type='snr',  $period='day'));
}

function rrd_hourly($freebox){
	# $f = 'rrd-rate-hourly-up.png';   file_rotate("$f", 24); file_put_contents("var/$f", $freebox->rrd_graph($type='rate', $period='hour', $dir='up'));
	$f = 'rrd-rate-hourly-down.png'; file_rotate("$f", 24); file_put_contents("var/$f", $freebox->rrd_graph($type='rate', $period='hour', $dir='down'));
	# $f = 'rrd-snr-hourly.png';       file_rotate("$f", 24); file_put_contents("var/$f", $freebox->rrd_graph($type='snr',  $period='hour'));

	# $f = 'rrd-wan-rate-hourly-up.png';   file_rotate("$f", 24); file_put_contents("var/$f", $freebox->rrd_graph($type='wan-rate', $period='hour', $dir='up'));
	# $f = 'rrd-wan-rate-hourly-down.png'; file_rotate("$f", 24); file_put_contents("var/$f", $freebox->rrd_graph($type='wan-rate', $period='hour', $dir='down'));

}

switch(get_arg(1, 'daily')){


	case 'get-ejs':

	$list = <<<EOF
		tpl/conn_ddns_provider_status.ejs
		tpl/fw_wan_redir.ejs
		tpl/igd_redir.ejs
		tpl/lfilter_entry.ejs
		tpl/nas_storage_disk_advanced_informations.ejs
		tpl/nas_storage_disk.ejs
		tpl/nas_storage_disk_format_config.ejs
		tpl/nas_storage_disk_format_config_simple.ejs
		tpl/nas_storage_disk_format_confirm.ejs
		tpl/nas_storage_disk_format_error.ejs
		tpl/nas_storage_disk_format_progress.ejs
		tpl/nas_storage_disk_format_success.ejs
		tpl/nas_storage_format_confirm.ejs
		tpl/nas_storage_format_error.ejs
		tpl/nas_storage_format_progress.ejs
		tpl/nas_storage_format_success.ejs
		tpl/nas_storage_partition_fsck_config.ejs
		tpl/nas_storage_partition_fsck_error.ejs
		tpl/nas_storage_partition_fsck_progress.ejs
		tpl/nas_storage_partition_fsck_result.ejs
		tpl/nas_storage_partition_fsck_unsupported.ejs
		tpl/net_ethsw_mac_address_table_entry.ejs
		tpl/net_ethsw_port_state.ejs
		tpl/net_ethsw_port_stats_counters.ejs
EOF;
		$list = explode("\n", $list);
		foreach($list as $f){
			$f = trim($f);
			printf("%s\n", $f);
			$content = $freebox->uri_get($f);
			print_r($content);
		}

		break;

	case 'hour': 
	case 'hourly': 
		rrd_hourly($freebox);
		break;

	case 'day': 
	case 'daily': 
		rrd_dayly($freebox);
		break;

	case 'week': 
	case 'weekly': 
		rrd_weekly($freebox);
		break;


	case 'json':
	case 'json-exec':
	case 'exec':
		print_r($freebox->exec('conn.status'));
		break;

	case 'dhcp': 
		print_r($cnf = $freebox->dhcp->config_get());
		# print_r($cnf = $freebox->dhcp->config_set($cnf));
		print_r($freebox->dhcp->status_get());
		print_r($freebox->dhcp->leases_get());
		print_r($freebox->dhcp->sleases_get());
		break;

	case 'system': 
		print_r($freebox->system->uptime_get());
		print_r($freebox->system->mac_address_get());
		print_r($freebox->system->fw_release_get());
		# print_r($freebox->phone->reboot($timeout=60));
		break;

	case 'phone': 
		# print_r($freebox->phone->fxs_ring(true)); sleep (3);
		# var_dump($freebox->phone->fxs_ring(false));
		print_r($freebox->phone->status());
		break;

	case 'ftp': 
		print_r($cnf = $freebox->ftp->get_config());
		# $cnf['enabled'] = 1;
		# $cnf['allow_anonymous'] = 1;
		# $cnf['allow_anonymous_write'] = 1;
		# $cnf['password'] = '123456';
		# $cnf = $freebox->ftp->get_config($cnf);
		break;

	case 'storage': 
		print_r($cnf = $freebox->storage->_list());
		print_r($cnf = $freebox->storage->disk_advanced_informations_get(0));
		# print_r($cnf = $freebox->storage->mount(1001));
		# print_r($cnf = $freebox->storage->partition_fsck(1001));
		# sleep (2) ; print_r($cnf = $freebox->storage->_list());
		# print_r($cnf = $freebox->storage->disable(1000));

		# print_r($cnf = $freebox->storage->disk_get(0));
		# print_r($cnf = $freebox->storage->disk_get(1));  # exception JSON
		# print_r($cnf = $freebox->storage->disk_get(1001)); # exception JSON
		# print_r($cnf = $freebox->storage->disk_get(1000)); # exception JSON
		# print_r($cnf = $freebox->storage->partition_simple(1000)); # exception JSON : method not found.

		break;

	case 'download': 

		print_r($freebox->download->_list());
		print_r($freebox->download->config_get());

		$args = array(
			'max_up' => 90,
			'download_dir' => '/Disque dur/Téléchargements',
			'max_dl' => 2234,
			'max_peer' => 240,
			'seed_ratio' => 2
		);
		# print_r($freebox->download->http_add('ubuntu.iso', 'ftp://ftp.free.fr/mirrors/ftp.ubuntu.com/releases/12.10/ubuntu-12.10-desktop-i386.iso'));
		# print_r($freebox->download->_list());
		$id = 849;
		$type='http';
		# print_r($freebox->download->stop($type, $id));
		# print_r($freebox->download->start($type, $id));
		# print_r($freebox->download->remove($type, $id));
		$url = 'http://ftp.free.fr/mirrors/ftp.ubuntu.com/releases/12.10/ubuntu-12.10-desktop-armhf%2bomap4.img.torrent';
		# print_r($freebox->download->torrent_add($url));

		# print_r($freebox->download->config_get());
		# print_r($freebox->download->config_set($args));
		break;

	case 'conn':
		# print_r($freebox->conn->status());
		var_dump($freebox->conn->remote_access_get());
		# var_dump($freebox->conn->remote_access_set(false));
		break;

	case 'fs':
		print_r($freebox->fs->_list('/Disque dur/Marc'));
		# print_r($freebox->fs->_list('/Disque dur/Maxime', 'with_attr'));
		# print_r($freebox->fs->_list('/', array('with_attr' => true)));

		# $opts = array('with_attr' => 1);
		# $args = array(
		# 	'/Disque dur/',
		# 	$opts
		# );
		# print_r($freebox->exec('fs.list', $args));
		break;

	case 'test': 
	case 'debug': 

		# ethsw.port_state
		# ethsw.mac_address_table
		# ethsw.port_counters
		# ethsw.port_set_config
		print_r($freebox->exec('ethsw.port_state'));

		# $freebox->debug();
		# print_r($cfg = $freebox->ftp->get_config());
		# print_r($freebox->ftp->set_config($cfg));

		break;
}


?>

File: ruby/test-mafreebox.rb
#!/usr/bin/ruby
# coding: utf-8

# author : Marc Quinton, février 2013, licence : http://fr.wikipedia.org/wiki/WTFPL

require 'pp'
require './Mafreebox.rb'

def usage
	puts "usage : ruby test-mafreebox [module]\n"
	puts "- permet de tester l'usage et le fonctionnement des modules\n"
	puts "- liste des modules supportés : system, download, fs, unix\n"
end


cnf= ENV['HOME'] + '/.config/mafreebox.yml'

if(File.exist?(cnf))
	config = YAML::load(File.open(cnf))

	# vous pouvez creer un fichier $HOME/.config/mafreebox.yml avec le contenu suivant :
	# :url: http://mafreebox.free.fr/ (ou adresse IP externe)
	# :login: freebox
	# :passwd: votre-mot-de-passe

else
	config = YAML::load(YAML::dump(
		:url		=> 'http://mafreebox.freebox.fr/', # ou votre adresse IP externe
		:login		=> 'freebox',
		:passwd		=> 'your-password'
	))

end

(usage && exit(0)) if ARGV.length == 0

begin
	mafreebox = Mafreebox::Mafreebox.new(config)
	# throws an exception on login or connexion error
	mafreebox.login
#rescue
#	puts "connexion error :"
#	puts "- url : " + config[:url]
#	puts "- login : " + config[:login]
#	puts "- passwd : " + config[:passwd]
#	exit(-1)
end

case ARGV[0]
	when "json"
		# on récupère des enregistrements de cette forme : 
		# [{"type"=>"dir", "name"=>"Directory-name", "mimetype"=>"type-mime"}
		mafreebox.exec('fs.list', ['Disque dur/']).each { |e|
			printf("%s : %s : %s\n", e['type'], e['name'], e['mimetype'])
		}

	when "system"
		pp mafreebox.system.get
		# pp mafreebox.system.reboot
		
	when "conn"
		pp mafreebox.conn.get
		
	when "phone"
		pp mafreebox.phone.logs

	when "ipv6"
		pp mafreebox.ipv6.get

	when "extra"
		version = mafreebox.system.fw_release_get
		puts "# firmware : #{version}"
		puts mafreebox.export_csv(mafreebox.extra.legal)

	when "igd"
		pp mafreebox.igd.get

	when "blink"
		# mafreebox.lcd.blink
		mafreebox.lcd.blink_code(5)

	when "lcd"
		if ARGV.count == 2
			pp mafreebox.lcd.set(ARGV[1])
		else	
			pp mafreebox.lcd.get
		end

	when "share"
		pp mafreebox.share.get

	when "user"
		pp mafreebox.user.password_check_quality("123")

	when "lan"
		pp mafreebox.lan.get
		
	when "wifi"
		pp mafreebox.wifi.get

		if(ARGV[1] == 'set-active')
			pp ARGV
			case ARGV[2]
				when true
				when "true"
				when "on"
					puts "set-active = on"
					mafreebox.wifi.set_active(true)
				when false
				when "false"
				when "off"
					puts "set-active = off"
					mafreebox.wifi.set_active(false)
			end
		else
			pp mafreebox.wifi.get
		end
	when "download"
		p mafreebox.download.list

		# sleep(1) ; p mafreebox.download.config_get
		# pp mafreebox.download.http_add('ubuntu-12.10-server-amd64.iso', 'http://www-ftp.lip6.fr/pub/linux/distributions/Ubuntu/releases/12.10/ubuntu-12.10-server-amd64.iso')
		# sleep 1 ; pp mafreebox.download.list

		pp mafreebox.download.config_get
		
		config = {
			'download_dir'=>'/Disque dur/Téléchargements',
			'seed_ratio'  =>4,
			'max_peer'    =>30,
			'max_dl'      =>30,
			'max_up'      =>20
		}
		# FIXME : ne fonctionne pas comme escompté ....
		# pp mafreebox.download.config_set(config)


	when "unix"
		mafreebox.unix.ls('/Disque dur/').each { |e| printf("- %s\n", e['name'])} ; puts("\n")

		# scénario de test : 
		# - liste /Disque dur/test
		# - depuis /Disque dur/test ; 
		#   - copie toto.txt titi.txt cp(toto.txt", "titi.txt")
		#   - renome toto.txt -> mv
		#   - efface tutu.txt -> rm
		#   - crée un répertoire, le renomme et l'efface (mkdir, mv, rmdir)

		#	mafreebox.unix.ls('/Disque dur/test').each { |e| printf("- %s\n", e['name'])} ; puts("\n")
		#	mafreebox.unix.cp('/Disque dur/test/toto.txt','/Disque dur/test/titi.txt')
		#	sleep(1); mafreebox.unix.ls('/Disque dur/test').each { |e| printf("- %s\n", e['name'])} ; puts("\n")

		#	mafreebox.unix.mv('/Disque dur/test/titi.txt','/Disque dur/test/tutu.txt')
		#	sleep(1); mafreebox.unix.ls('/Disque dur/test').each { |e| printf("- %s\n", e['name'])} ; puts("\n")

		#	mafreebox.unix.rm('/Disque dur/test/tutu.txt')
		#	sleep(1); mafreebox.unix.ls('/Disque dur/test').each { |e| printf("- %s\n", e['name'])} ; puts("\n")

		#	mafreebox.unix.mkdir('/Disque dur/test/foobar')
		#	sleep(1); mafreebox.unix.ls('/Disque dur/test').each { |e| printf("- %s\n", e['name'])} ; puts("\n")

		#	mafreebox.unix.mv('/Disque dur/test/foobar', '/Disque dur/test/bar')
		#	sleep(1); mafreebox.unix.ls('/Disque dur/test').each { |e| printf("- %s\n", e['name'])} ; puts("\n")

		#	mafreebox.unix.rmdir('/Disque dur/test/bar')
		#	sleep(1); mafreebox.unix.ls('/Disque dur/test').each { |e| printf("- %s\n", e['name'])} ; puts("\n")

	when "fs"
		# p mafreebox.fs.list('/')
		# p mafreebox.fs.list('/Disque dur/')
		# p mafreebox.fs.json_get('/Disque dur/test/toto.txt') # FIXME : erreur !

		# require 'digest'
		# content = mafreebox.fs.get('/Disque dur/test/toto.txt')
		# printf("%s", Digest::MD5.hexdigest(content))

		# p mafreebox.fs.copy('/Disque dur/test/toto.txt', '/Disque dur/test/titi.txt')
		# p mafreebox.fs.remove('/Disque dur/test/titi.txt')
		# p mafreebox.fs.mkdir('/Disque dur/test/foobar')
		# p mafreebox.fs.move('/Disque dur/test/foobar', '/Disque dur/test/tutu')

end


